core/src/main/java/io/questdb/griffin/model/RuntimeIntervalModel.java

```diff
@@ -0,0 +1,146 @@
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.griffin.model;
+
+import io.questdb.cairo.sql.Function;
+import io.questdb.griffin.SqlExecutionContext;
+import io.questdb.std.LongList;
+import io.questdb.std.ObjList;
+import io.questdb.std.datetime.microtime.Timestamps;
+
+import static io.questdb.griffin.model.IntervalUtils.STATIC_LONGS_PER_DYNAMIC_INTERVAL;
+import static io.questdb.griffin.model.RuntimeIntervalModelBuilder.DYNAMIC_LO_HI;
+
+public class RuntimeIntervalModel implements RuntimeIntrinsicIntervalModel {
+    // These 2 are incoming model
+    private final LongList intervals;
+    private final ObjList<Function> dynamicRangeList;
+
+    // This used to assemble result
+    private LongList outIntervals;
+
+    public RuntimeIntervalModel(LongList intervals) {
+        this(intervals, null);
+    }
+
+    public RuntimeIntervalModel(LongList staticIntervals, ObjList<Function> dynamicRangeList) {
+        this.intervals = staticIntervals;
+
+        this.dynamicRangeList = dynamicRangeList;
+    }
+
+    @Override
+    public LongList calculateIntervals(SqlExecutionContext sqlContext) {
+        if (isStatic()) {
+            return intervals;
+        }
+
+        if (outIntervals == null) {
+            outIntervals = new LongList();
+        } else {
+            outIntervals.clear();
+        }
+
+        // Copy static part
+        int dynamicStart = intervals.size() - dynamicRangeList.size() * STATIC_LONGS_PER_DYNAMIC_INTERVAL;
+        outIntervals.add(intervals, 0, dynamicStart);
+
+        // Evaluate intervals involving functions
+        addEvaluateDynamicIntervals(outIntervals, sqlContext);
+        return outIntervals;
+    }
+
+    @Override
+    public boolean isFocused(Timestamps.TimestampFloorMethod floorMethod) {
+        if (!isStatic()) return false;
+
+        long floor = floorMethod.floor(intervals.getQuick(0));
+        for (int i = 1, n = intervals.size(); i < n; i++) {
+            if (floor != floorMethod.floor(intervals.getQuick(i))) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private void addEvaluateDynamicIntervals(LongList outIntervals, SqlExecutionContext sqlContext) {
+        int dynamicStart = intervals.size() - dynamicRangeList.size() * STATIC_LONGS_PER_DYNAMIC_INTERVAL;
+        int dynamicIndex = 0;
+
+        for (int i = dynamicStart; i < intervals.size(); i += STATIC_LONGS_PER_DYNAMIC_INTERVAL) {
```

Comment:
could be slightly faster loop with pre-calculated size

Reply:
ok